pypicloud/cache/base.py

```diff
@@ -170,6 +172,71 @@ def distinct(self):
         """
         raise NotImplementedError
 
+    def search(self, criteria, query_type):
+        """
+        Perform a search from pip
+
+        Parameters
+        ----------
+        criteria : dict
+            Dictionary containing the search criteria. Pip sends search criteria
+            for "name" and "summary" (typically, both of these lists have the
+            same search values).
+
+            Example:
+            {
+                "name": ["value1", "value2", ..., "valueN"],
+                "summary": ["value1", "value2", ..., "valueN"]
+            }
+
+        query_type : str
+            Type of query to perform. By default, pip sends "or".
+
+        """
+        name_queries = criteria.get('name', [])
+        summary_queries = criteria.get('summary', [])
+        packages = []
+        packages_found = set()
+
+        for key in self.distinct():
+            # Search all versions of this package key
+            for package in self.all(key):
+                # Skip this result if it has already been selected
+                if package.name in packages_found:
+                    continue
+
+                # Search package names
+                for query in name_queries:
+                    # Look for this query anywhere in the package name
+                    if query.lower() in package.name.lower():
+                        # Found a match, adding to the packages_found
+                        # set and generating a result
+                        packages_found.add(package.name)
+                        packages.append({
+                            'name': package.name,
+                            'summary': package.summary,
+                            'version': package.version,
+                        })
```

Comment:
I think that here it makes more sense to simply append the `package` directly. I think the only place the return value is used right now is in the search endpoint, and if you return the python objects directly they will be serialized with the `__json__()` method. That will add the additional `last_modified` and `url` fields, but if those don't cause problems for pip then that might be a cleaner way to do it.